### DISCUSSION TITLE
Add extra stepper driver to stock board for proper dual z ?

### DISCUSSION THREAD

**Discussion by Colin66** [bassamanator/everything-sovol-sv06#37] [Ideas]
Hi, do you think there's any way to add a 5th stepper driver to the SV06/+ board ?  
Having proper independent Z levelling would be a great improvement. My electronics knowledge isn't sufficient but it appears to be an ender clone board and I've seen people accessing unused pins either directly from the processor or via a header : 
standalone mode :
https://www.reddit.com/r/ender3v2/comments/qzxsok/i_added_a_5th_stepper_driver_to_the_stock/
https://docs.google.com/document/d/1O6iDR2msre3tUxXAF8RXY9PUIZHn8q9FAH5B1m9y7AY/edit?pli=1
UART mode :
https://www.reddit.com/r/ender3v2/comments/n3clew/ender_3v2_422427_board_tmc_uart_mods/
https://drive.google.com/file/d/1oUj07YvqexBF21gjLlnnvFgyJlPtkhSL/view?pli=1

**Comment by bassamanator** [comment]
Very cool! Looks like it might be possible based purely on the similarity of appearance of the boards.

For the record, as long as your machine is square, gantry levelling is a non issue. I think I levelled by gantry 5 or 6 months ago. I even switched out my bed for a new one about 3 weeks ago, and I still didn't level the gantry, complete non issue.

Unfortunately, I really don't know about the hard technicalities here. There's a user on here that knows quite a bit about this stuff, maybe he'll chime in. @BeauSlim

**Comment by occasional-contributor** [comment]
If you are already running Klipper, you could get a CAN tool board like EBB42 for the extruder, then reuse the extruder driver on the main board for the second z axis.

**Reply by Colin66** [reply]
Thanks, I didn't know about that. I guess it would be a bit of a carve up with a board needing to be accommodated on the print head.  It would be nice to just mod the main board and run a wire to the other z motor if poss.

**Reply by occasional-contributor** [reply]
I replaced the main board with an SKR Pico and EBB36 just for learning. Eventually, I want to build a CoreXY printer from scratch. Voron 0 or a smaller Voron 2 or something similar.

**Reply by occasional-contributor** [reply]
1. You do not need to put the tool board in the extruder. If you hotend is completely stock, you can disconnect the harness from the main board and hook up the loose ends to the tool board instead.
2. The EBB42 can be used over USB or CAN. You can choose which you want to use when building the Klipper firmware.

**Reply by indistinctTalk** [reply]
In a similar vein, I've used the Fysetc ERB board to add additional drivers at a relatively cheap price point.

**Comment by Heiko910** [comment]
Very interesting ideas here.
I would love to get either a driver freed up to use it for the second z
axis stepper or adding a drive. Either way would be great.

**Comment by BeauSlim** [comment]
The mod described for the Creality board should work.  Those are the same 2 pins I used in my [fan MOSFET mod](https://www.printables.com/model/672213-sovol-sv06plus-fan-mod-with-mosfet-lid-for-mcu-enc/files)    

I was planning on doing dual Z on my SV06 but anti-backlash nuts do a decent job of keeping my Z screws from going out of sync while powered down.  Adding a toolhead board to the MCU case is an interesting idea, but I think you can get a full 5-motor board with drivers for less than double the price.

**Reply by occasional-contributor** [reply]
For sure. The only reason I went with SKR Pico + EBB42 is because I am building a custom 3D printer and I got these at deeply discounted prices during Black Friday.

**Comment by Colin66** [comment]
Looking at the first mod (the 2 header pins and standalone mode), it looks like the motor is wired permanently on, in which case I guess you'd be unable to disable steppers from the touch pad and manually adjust the z by hand, although you probably never need to.
Is there any problem having one z driver in full UART mode and the other in standalone dumb mode or is this just a per-driver setting in the firmware ?

Full UART mode looks better for sure but there's a risk of damage soldering to those tiny pins on the ic.
It would be so much better if the manufacturers led unused pins to solder pads or a header..

**Reply by BeauSlim** [reply]
Yes, you can just configure the on-board Z driver in standalone mode.  Sovol's klipper.cfg for their KlipperScreen unit does this anyway.

**Comment by Ederhex** [comment]
I've been experimenting with an ebb42 since yesterday. Had some issues with the configuration but as of now I'm starting my first print with the extra stepper driver. I will post my findings when I'm confident everything is working fine.

**Reply by BeauSlim** [reply]
Definitely let us know.  Just finishing up an order from AliExpress including an ebb board.  Was thinking of going with the 36 and putting it on the toolhead on my SV06 Plus.  Are you planning something similar or putting the EBB in the MCU box?

**Comment by dasb00ter** [comment]
> I've been experimenting with an ebb42 since yesterday. Had some issues with the configuration but as of now I'm starting my first print with the extra stepper driver. I will post my findings when I'm confident everything is working fine.

Well how did it go... Are u seriously injured by flying gantry debris... Should we send help?

**Reply by Colin66** [reply]
That's a coincidence, I've just today managed to add an extra stepper driver connecting directly to the main board and now have z_tilt running successfully.   
I'll write a quick guide re how to connect it up.

**Comment by Colin66** [comment]
OK so I figured it out and have dual Z drivers working now so can do Z_Tilt.
Did a bit of a write-up here : 
https://github.com/Colin66/Add-extra-stepper-to-STM32F103-32-bit-boards-including-Sovol-SV06

**Reply by BeauSlim** [reply]
Great job finding the extra pin.  There are so few available on that board.

**Comment by dasb00ter** [comment]
> OK so I figured it out and have dual Z drivers working now so can do Z_Tilt. Did a bit of a write-up here : https://github.com/Colin66/Add-extra-stepper-to-STM32F103-32-bit-boards-including-Sovol-SV06

This is a pretty comprehensive write up actually. It looks like a 5 dollar part off of AliExpress. Just the labour and time. In case I actually get the time to do this what sort of guage and type of wire for this? I do some soldering boards but mostly home automation esp32 boards and such (5v's). Maybe @bassamanator knows more about the motor settings. If I get past not letting out the magic smoke out of my main board I might as well avoid burning the motor out. I noticed that allot of hey let's throw klipper on our sv06 (plus) stream videos resulted in hot motors. So maybe that's just settings

**Reply by Colin66** [reply]
Thanks. Most of the cables I used were standard stepper cables (not sure of the gauge), except for the Dupont jumpers I used to reconnect the filament sensor to the z- limit input.  You can actually use both ends of a stepper cables, one to the motor and one from the 4 pin header. I think the tmc driver modules are about £4.  An easy but worthwhile upgrade.

**Reply by dasb00ter** [reply]
I've ordered the parts, so anywhere from 1 wk to 4 months waiting lol. Hopefully I will get some time too. In the off chance you have a photo or have to pull parts to look would it be possible to see what the other side of your perf board looks like. I havent used prototype board but I should really start and I am curious about how you have made this better for deconstruction if necessary.
Here is my BOM for the project:
https://www.aliexpress.com/item/1005005945738204.html
https://www.aliexpress.com/item/1005006172387930.html
https://www.aliexpress.com/item/4000947676742.html
https://www.aliexpress.com/item/32809917073.html

**Comment by dasb00ter** [comment]
@Colin66 looking at your diagram and the board you soldered I don't see some connections. Are they on the reverse of the board? I see 6 connections along the uart side in the diagram but you soldered board appears different. What's going on with the green line in the diagram? I have the same parts and a SV06 + and I'm adventurous and can solder but I don't understand electronic circuits that well. Thanks for this though

**Reply by Colin66** [reply]
Hi there, yes, the connections are on the reverse side.  Here's a pic, but my soldering is pretty messy, I'm going with the excuse that my soldering iron wasn't working properly :).  T
![IMG_20240520_121103747](https://github.com/bassamanator/everything-sovol-sv06/assets/45802669/95f5ab2e-1463-4c41-94ec-105645f85784)
The green line from the 4 pin header on the main board next to the screen port is the 0V line for the logic side of the board and connects to the driver board GND next to the VCC pin and also to the pin marked EN on the driver board.  The other wire I've got marked green is just one of the 4 stepper coil wires.    The other 0V line which provides a return for the 24V powering the motor (which one leg of the capacitor connects to), I've shown in black, and it goes to the GND pin on the bottom left next to the 24V feed and the small potentiometer.  I'd definitely keep that 0V separate from the green one i described above.
The capacitor will explode if you connect it the wrong way around, it will probably have a bunch of - symbols ----- marked on the heatshrink sleeve.  that goes to the GND pin.  Here's a clear pic of the stepper pinouts :
![ND1172_4-min-min](https://github.com/bassamanator/everything-sovol-sv06/assets/45802669/833beab4-414f-4936-9a4c-8d55ac9bf056)

**Reply by dasb00ter** [reply]
Thanks for your quick reply. 
If it's ugly and it works leave it alone is my experience. 
U may want to consider putting that pic on your GitHub... More people like me may come along. 
I've made sure to get that same board so I'm just really gonna copy cat what you did. 
Were you able to reduce the heat generated on the motor?

**Reply by Colin66** [reply]
Once you've got it all built you'll want to double and triple check all the wiring and be very tentative about powering the stepper with it just in case you have something configured wrongly and only one motor moves or the two move in opposite directions which would damage your X gantry.  You can force a small motor move without homing by entering SET_KINEMATIC_POSITION then G0 Z0.1 and check they both move and go in the same direction.  I replaced the left motor drive and my config is as in my github repo.  Keep an eye on your first Z_TILT_ADJUST because if you have that set up backwards it'll keep trying to adjust but will get worse each time, putting the gantry more and more out of alignment.
Mine is working 100% fine and it auto aligns at the start of each print, integrated into the KAMP routine.

**Reply by Colin66** [reply]
Re motor heat, yes, i just reduced the run and hold currents.  The hold torque is a bit lower as a consequence and there will probably be a better way to do it using the low level TMC commands but there are so many of those I don't know which to set.  
You can dump the list of TMC parameters from any of the drivers using DUMP_TMC STEPPER=stepper_XX so can see the list of settings for the on-board drivers, I guess they're the optimal settings.  Mine just works and I haven't  played around further.  I was a bit reluctant to add the pic not just to avoid the solder shame but because the driver pinouts do vary between different versions of the boards and a simple copy might be wrong.

**Reply by dasb00ter** [reply]
> Once you've got it all built you'll want to double and triple check all the wiring and be very tentative about powering the stepper with it just in case you have something configured wrongly and only one motor moves or the two move in opposite directions which would damage your X gantry. You can force a small motor move without homing by entering SET_KINEMATIC_POSITION then G0 Z0.1 and check they both move and go in the same direction. I replaced the left motor drive and my config is as in my github repo. Keep an eye on your first Z_TILT_ADJUST because if you have that set up backwards it'll keep trying to adjust but will get worse each time, putting the gantry more and more out of alignment. Mine is working 100% fine and it auto aligns at the start of each print, integrated into the KAMP routine.

You know for this part I thought I would just remove the x gantry and watch the screws. I mean if your digging around in your control box and adding components seems like should be in the ballpark. I suppose some might be confident enough to not do that though

**Comment by BeauSlim** [comment]
Just commenting to say that dual Z is worth doing.  I added an EBB36 to my SV06 Plus toolhead (driving the E motor, etc.) and rewired the "E" driver from the main MCU board to the second Z motor.  Z_TILT_ADJUST makes my X so parallel to the bed that I'm getting the best first layers I've ever seen.  
Thanks to @Colin66 for the config/macro.

**Comment by indistinctTalk** [comment]
Yeah, it doesn't fix a bad bed but it does resolve a problem that can
contribute.
